### PR TITLE
parameter for parameter autosave

### PR DESCRIPTION
**Pull request by thomasgubler**
This is a follow up on https://github.com/PX4/Firmware/issues/1903 and https://github.com/PX4/Firmware/commit/f8fd67d3e17e8f32b7ee5a7d9e112bf944e2735f

Adding a parameter to control the parameter autosave functionality. Turn autosave off by default as it is not according to mavlink specifications. If I remember correctly we talked about the same "one save button" option for qgroundcontrol. However the GCS should adhere to the protocol and send 2 commands (set/write).

The autosave functionality is a good workaround as long as the Tower app does not support the parameter feature correctly.

tested with qgc and command line

### CODE REVIEW

**Comment by LorenzMeier** [comment]
Could you explain the value why we want this? I agree on the interpretation of the mavlink spec, but in reality nobody does use non-sticky parameters.
The two-step saving caused huge usability issues for our users, so I would not recommend the change in the Tower app either.
Modern systems and UIs tend to store changes immediately and put them into effect straight ahead.

**Comment by thomasgubler** [comment]
> but in reality nobody does use non-sticky parameters.

I know several people (developers) who like to be able to play around and go back to sane parameters quickly.

> The two-step saving caused huge usability issues for our users, so I would not recommend the change in the Tower app either.
Modern systems and UIs tend to store changes immediately and put them into effect straight ahead.

I am not against this change on the UI side. Actually I think that for the same reasons you mention it should be the default option in QGC and it certainly makes sense for an app like Tower. However nothing speaks against letting the UI and the developer of that software decide about the user interaction. If he wants to use the one step solution he can just adhere to the protocol and silently send both mavlink commands. This seems more flexible and leaves for example room for an "advanced mode" where you have the set/write options. For me it seems that the autosave functionality unnecessarily tries to solve UX related problems on the Firmware level.

The autosave implementation is also to some degree a hack on top of the parameter interface. If we don't support non-persistent changes why not disable the functionality in the parameter interface instead of the implementation in the commander?

**Comment by LorenzMeier** [comment]
Because some subsystems like the calibration now rely on parameters being persistent. If you want to play with parameters why not store your sane set on the ground and re-upload them after done with playing? The double storage of parameters is still a huge source of confusion (in particular for developers, ask @tumbili or @kd0aij).
I'm trying to make the system fit for 99% of the daily use, and for the 1% of exceptions it seems doable to store a copy on the ground before doing any editing. Upon re-loading the file the param UI will also show the diff between the ones onboard and the re-loaded ones from file.

**Comment by DonLakeFlyer** [comment]
My plan was to either get rid of the two step in QGC parameter editor all together. Or hide it somehow in a deep dark place. All other internal QGC code sets the param to rom immediately.

**Comment by thomasgubler** [comment]
>  Because some subsystems like the calibration now rely on parameters being persistent.

To me this seems to imply that we can remove non-persistent parameters altogether. Why do we need to rely on the autosave loop in the commander?  Why can't those subsystems use a param save call (either issued by the UI or self-issued directly after the calibration) or a *persistent* param set? 

>If you want to play with parameters why not store your sane set on the ground and re-upload them after done with playing?

My experience tells me that with cold fingers powercycling the system is much easier than fiddling around with parameter files. Also because you power-cycle automatically during flashing the automatic parameter reset makes development much easier: e.g. you can flash, change parameters, observe a problem, adapt your code, flash and restart at the 'default' state without the need to connect to qgc, browse to a param file and finally upload those params.

>My plan was to either get rid of the two step in QGC parameter editor all together. Or hide it somehow in a deep dark place. All other internal QGC code sets the param to rom immediately.

I completely agree with this. This demonstrates that you can do a good GUI/UX using the set/write commands on the firmware level. The whole process is completely hidden from the user. Why not make the system fit for 100% of the use cases by doing the change in the user interface?

In summary I don't quite get if you @LorenzMeier  are trying to get rid of non-persistent parameters altogether or not. The current solution still offers all the API/specs for non-persistent storage but silently makes everything persistent through autosave.

**Comment by SimonWilks** [comment]
I can understand the need to default to autosave on for the sake of the average user but would really (really) appreciate the ability to disable this at least for tuning params. Obviously calibration params can be autosaved.

Seeing this is not the first time this 99:1 topic has come up (and may belong somewhere else), while I agree on focusing on the 99% of users those 1% help keep things running and should be credited with a little more than such a low single digit figure in the whole equation.

**Comment by DonLakeFlyer** [comment]
The whole autosave thing seems odd to me. Not sure why the GCS shouldn't handle that for you and provide you with the option to do what you want (appropriately exposed, to power users).  If the GCS is implemented incorrectly to the mavlink spec, fix the GCS.

**Comment by LorenzMeier** [comment]
Re-kicked the CI systems.

**Comment by LorenzMeier** [comment]
It sounds like this topic warrants a wider discussion - is it maybe time to establish a bi-weekly dev call?

**Comment by LorenzMeier** [comment]
I think we will cover this topic on the upcoming dev call, for now I've rebased and applied this, but defaulted it to on. Developers can still set it to off and we will work towards having all GCS comply with the MAVLink spec. I will make sure that other parts of the code do not silently assume autosave (like the calibration routines, which currently do).
